crates/bevy_asset/src/path.rs

```diff
@@ -67,38 +75,128 @@ impl<'a> Display for AssetPath<'a> {
     }
 }
 
+#[derive(Error, Debug, PartialEq, Eq)]
+pub enum ParseAssetPathError {
+    #[error("Asset source must be followed by '://'")]
+    InvalidSourceSyntax,
+    #[error("Asset source must be at least one character. Either specify the source before the '://' or remove the `://`")]
+    MissingSource,
+    #[error("Asset label must be at least one character. Either specify the label after the '#' or remove the '#'")]
+    MissingLabel,
+}
+
 impl<'a> AssetPath<'a> {
     /// Creates a new [`AssetPath`] from a string in the asset path format:
     /// * An asset at the root: `"scene.gltf"`
     /// * An asset nested in some folders: `"some/path/scene.gltf"`
     /// * An asset with a "label": `"some/path/scene.gltf#Mesh0"`
+    /// * An asset with a custom "source": `"custom://some/path/scene.gltf#Mesh0"`
     ///
     /// Prefer [`From<'static str>`] for static strings, as this will prevent allocations
     /// and reference counting for [`AssetPath::into_owned`].
-    pub fn new(asset_path: &'a str) -> AssetPath<'a> {
-        let (path, label) = Self::get_parts(asset_path);
-        Self {
+    ///
+    /// # Panics
+    /// Panics if the asset path is in an invalid format. Use [`AssetPath::try_parse`] for a fallible variant
+    pub fn parse(asset_path: &'a str) -> AssetPath<'a> {
+        Self::try_parse(asset_path).unwrap()
+    }
+
+    /// Creates a new [`AssetPath`] from a string in the asset path format:
+    /// * An asset at the root: `"scene.gltf"`
+    /// * An asset nested in some folders: `"some/path/scene.gltf"`
+    /// * An asset with a "label": `"some/path/scene.gltf#Mesh0"`
+    /// * An asset with a custom "source": `"custom://some/path/scene.gltf#Mesh0"`
+    ///
+    /// Prefer [`From<'static str>`] for static strings, as this will prevent allocations
+    /// and reference counting for [`AssetPath::into_owned`].
+    ///
+    /// This will return a [`ParseAssetPathError`] if `asset_path` is in an invalid format.
+    pub fn try_parse(asset_path: &'a str) -> Result<AssetPath<'a>, ParseAssetPathError> {
```

Comment:
Would this benefit from having a corresponding `try_parse_static` function to make use of the `CowArc::Static` variant?

Reply:
Hmm yeah thats probably a good idea. We have `AssetPath::from(static_str)` for static strs. Maybe we should impl `TryFrom` instead for parity?